logics/impls/controller/mod.rs

```diff
@@ -713,33 +719,69 @@ impl<T: Storage<Data>> Internal for T {
     ) -> Result<()> {
         Ok(()) // do nothing
     }
+
     default fn _redeem_allowed(
         &self,
         pool: AccountId,
         redeemer: AccountId,
         redeem_amount: Balance,
-        pool_attribute: Option<PoolAttributes>,
+        pool_attributes: Option<PoolAttributes>,
     ) -> Result<()> {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
+        let (
+            AccountCollateralData {
+                total_collateral_in_base_currency,
+                avg_ltv: _,
+                avg_liquidation_threshold,
+                total_debt_in_base_currency,
+                asset_price,
+                liquidation_threshold,
+                health_factor: _,
+            },
+            asset_params,
+        ) = self._calculate_user_account_data(redeemer, pool_attributes)?;
+
+        // Prepare parameters for calculation
+        let (sum_collateral, sum_borrow_plus_effect) =
+            get_hypothetical_account_liquidity(GetHypotheticalAccountLiquidityInput {
+                asset_params,
+                token_modify: Some(pool),
+                redeem_tokens: redeem_amount,
+                borrow_amount: 0,
+            });
+
+        // These are safe, as the underflow condition is checked first
+        let shortfall = if sum_collateral > sum_borrow_plus_effect {
+            U256::from(0)
         } else {
-            None
+            sum_borrow_plus_effect.sub(sum_collateral)
```

Comment:
I think I can summarize since the only code that uses shortfall is the one immediately after

```rust
        if sum_collateral < sum_borrow_plus_effect {
            return Err(Error::InsufficientLiquidity)
        }
```